README.md
# Unity GameJolt API
A GameJolt API wrapper for Unity

# Installation
See: [Installing from a Git URL](https://docs.unity3d.com/Manual/upm-ui-giturl.html)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Session system.
- Added a null check for the `queries` arugment in `APIRequest`.
- More error types for catching different types of error.
- Optional argument for `APIRequest` for specifying whether or not to automatically handle errors(useful in cases where the endpoint behavior for error is different).

### Changed
- Updated some internally used variable names.
- Update `UserData` and `TrophyData` to use enums where possible.

## 0.3.5 - 2021-04-10
### Changed
- All methods have been rewritten to be static and use async. This will make handling errors much easier and eliminate the need for all of the callbacks internally used.
- Organized methods into their respective sub-classes.
- Moved GET method in the helper class to a new package.

### Removed
- Basically all of the old coroutine-reliant methods.

## 0.3.4 - 2021-02-12
### Changed
- Some under-the-hood optimizations...*?*
    - The wrapper now internally uses `APIRequest()` instead of creating request URLs from scratch.
    - ~~removed 7.3943662 percent of the lines~~

## 0.3.3 - 2021-02-10
### Changed
- `Authenticate()`'s callback now gets invoked upon every authentication attempt and not just successful ones.

## 0.3.2 - 2021-02-10
### Added
- New `forced` bool parameter to force a re-authentication when using `Authenticate()`.
### Changed
- `Authenticate()` now do not authenticate if `isAuthenticated` is already true and
- `Authenticate()` also now blocks any authentication attempt if there is already an attempt going on.
- `isAuthenticating` is also available for checking the authentication process.
- `ConstructTrophy()` has been removed in favor of class constructor. The API now internally uses the constructor.

## 0.3.1 - 2021-02-09
### Changed
- Simplified TrophyData object initialization.

## 0.3.0 - 2021-02-09
### Added
- `GrantTrophy()` for granting trophies and
- `RevokeTrophy()` for revoking them.
- `FetchTrophy()` for getting informations about a trophy.
- `ListTrophies()` to get a list of trophies.
- `OnTrophy` event.
- Documented methods with summaries.

## 0.2.0 - 2021-02-08
### Added
- Basic user authentication.
- `OnAuthenticate` event.
- Username and token caching.
- Error check for API calls.

### Removed
- Redundant `Debug.Log()`s when making API calls.

## 0.1.0 - 2021-02-08
### Added
- First implementation of the `GetServerTime()` method.


Runtime/Classes.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using DateTime = System.DateTime;
using DateTimeOffset = System.DateTimeOffset;
using SimpleJSON;

namespace WattanaGaming.GameJoltAPI
{
    public class UserData
    {
        public int id;
        public string avatarURL;
        public string userName;
        public string displayName;
        public string description;
        public string website;
        public Type type;

        public string signedUp;
        public DateTime signedUpTime;
        public string lastLoggedIn;
        public DateTime lastLoggedInTime;

        public Status status;

        public UserData(JSONNode JSONData)
        {
            id = JSONData["id"];
            avatarURL = JSONData["avatar_url"];
            userName = JSONData["username"];
            displayName = JSONData["developer_name"];
            description = JSONData["developer_description"];
            website = JSONData["developer_website"];

            {
                string typeString = JSONData["type"];
                type = typeString switch
                {
                    "User" => Type.User,
                    "Developer" => Type.Developer,
                    "Moderator" => Type.Moderator,
                    "Administrator" => Type.Administrator,
                    _ => throw new APIError("Unknown user type.")
                };
            }

            signedUp = JSONData["signed_up"];
            signedUpTime = DateTimeOffset.FromUnixTimeSeconds(JSONData["signed_up_timestamp"].AsLong).DateTime;
            lastLoggedIn = JSONData["last_logged_in"];
            lastLoggedInTime = DateTimeOffset.FromUnixTimeSeconds(JSONData["last_logged_in_timestamp"].AsLong).DateTime;

            {
                string statusString = JSONData["status"];
                status = statusString switch
                {
                    "Active" => Status.Active,
                    "Banned" => Status.Banned,
                    _ => throw new APIError("Unknown user status.")
                };
            }
        }

        public enum Type
        {
            User,
            Developer,
            Moderator,
            Administrator
        }

        public enum Status
        {
            Active,
            Banned
        }
    }
    
    public class TrophyData
    {
        public int id;
        public string title;
        public Difficulty difficulty;
        public string description;
        public string imageURL;
        public string achieved;

        public TrophyData(JSONNode JSONData)
        {
            id = JSONData["id"];
            title = JSONData["title"];
            
            {
                string difficultyString = JSONData["difficulty"];
                difficulty = difficultyString switch
                {
                    "Bronze" => Difficulty.Bronze,
                    "Silver" => Difficulty.Silver,
                    "Gold" => Difficulty.Gold,
                    "Platinum" => Difficulty.Platinum,
                    _ => throw new APIError("Unknown trophy difficulty.")
                };
            }

            description = JSONData["description"];
            imageURL = JSONData["image_url"];
            achieved = JSONData["achieved"];
        }

        public enum Difficulty
        {
            Bronze,
            Silver,
            Gold,
            Platinum
        }

        public override string ToString()
        {
            return $"ID: {id}, Title: {title}, Difficulty: {difficulty}, Description: {description}, Achieved: {achieved}";
        }
    }

    public enum TrophyEventType
    {
        Grant,
        Revoke
    }

    public enum SessionStatus
    {
        Active,
        Idle
    }

    public enum APIVersion
    {
        v1_2
    }

    public class APIError : System.Exception
    {
        public APIError(string message) : base(message) { }
    }

    public class AuthError : System.Exception
    {
        public AuthError(string message) : base(message) { }
    }
}


Runtime/GameJolt.cs
using System.Collections;
using System.Collections.Generic;
using System.Threading.Tasks;
using UnityEngine;
using DateTime = System.DateTime;
using DateTimeOffset = System.DateTimeOffset;
using SimpleJSON;
using WattanaGaming.WebRequest;

namespace WattanaGaming.GameJoltAPI
{
    public static class GameJolt
    {
        public static string GameID;
        public static string GameKey;

        // User credentials are cached for later use after authentication.
        public static string UserName { get; private set; }
        public static string UserToken { get; private set; }

        public static bool IsAuthenticated { get; private set; }
        public static bool IsAuthenticating { get; private set; }

        public static APIVersion Version = APIVersion.v1_2;

        // Public events for obvious stuff.
        public static event System.Action OnAuthenticated;
        public static event System.Action<TrophyEventType, int> OnTrophy;

        public static class User
        {
            public static async Task Authenticate(string name, string token, bool forced = false)
            {
                if ((IsAuthenticated && !forced) || IsAuthenticating)
                {
                    throw new AuthError("Already authenticated or is currently authenticating.");
                }
                Debug.Log($"Attempting to authenticate as {name}...");
                IsAuthenticating = true;
                try
                {
                    await APIRequest("users/auth/", new string[] { $"username={name}", $"user_token={token}" });
                    UserName = name;
                    UserToken = token;
                    IsAuthenticated = true;
                    IsAuthenticating = false;
                    Debug.Log("Successfully authenticated.");
                    OnAuthenticated?.Invoke();
                }
                catch (AuthError)
                {
                    UserName = UserToken = "";
                    IsAuthenticated = false;
                    IsAuthenticating = false;
                    throw;
                }
            }

            public static async Task<UserData> Fetch(string user, bool id = false)
            {
                return (await Fetch(new string[] { user }, id))[0];
            }

            public static async Task<List<UserData>> Fetch(string[] users, bool id = false)
            {
                Debug.Log($"Fetching user data for {string.Join(",", users)}...");
                List<string> queries = new List<string>();
                if (id == true)
                {
                    queries.Add($"user_id={string.Join(",", users)}");
                }
                else
                {
                    queries.Add($"username={string.Join(",", users)}");
                }
                JSONArray userArray = (await APIRequest("users/", queries.ToArray()))["users"].AsArray;
                List<UserData> userDatas = new List<UserData>();
                if (userArray != null)
                {
                    foreach (JSONObject user in userArray)
                    {
                        userDatas.Add(new UserData(user));
                    }
                }
                Debug.Log("Fetched user data.");
                return userDatas;
            }
        }

        public static class Session
        {
            public static async Task Open()
            {
                RequireAuthenticated("Attempt to open a game session without an authenticated user.");
                Debug.Log("Opening a new game session...");
                await APIRequest("sessions/open/", new string[] { $"username={UserName}", $"user_token={UserToken}" });
                Debug.Log("Opened a new game session.");
            }

            public static async Task Ping(bool setStatus = false, SessionStatus status = SessionStatus.Idle)
            {
                RequireAuthenticated("Attempt to ping a game session without an authenticated user.");
                Debug.Log("Pinging the game session...");
                List<string> queries = new List<string>() { $"username={UserName}", $"user_token={UserToken}" };
                if (setStatus)
                {
                    queries.Add(status switch
                    {
                        SessionStatus.Active => "active",
                        SessionStatus.Idle => "idle",
                        _ => throw new APIError("Invalid session status.")
                    });
                }
                await APIRequest("sessions/ping/", queries.ToArray());
                Debug.Log("Pinged game session.");
            }

            public static async Task<bool> Check()
            {
                RequireAuthenticated("Attempt to check a game session without an authenticated user.");
                JSONNode response = await APIRequest("sessions/check/", new string[] { $"username={UserName}", $"user_token={UserToken}" }, false);
                if (response["message"]) { throw new APIError(response["message"]); }
                Debug.Log("Checked game session.");
                return response["success"];
            }

            public static async Task Close()
            {
                RequireAuthenticated("Attempt to close a game session without an authenticated user.");
                Debug.Log("Closing the game session...");
                await APIRequest("sessions/close/", new string[] { $"username={UserName}", $"user_token={UserToken}" });
                Debug.Log("Closed the game session.");
            }
        }

        public static class Trophy
        {
            public static async Task Grant(int id)
            {
                RequireAuthenticated("Attempt to grant trophy without an authenticated user.");
                Debug.Log($"Granting trophy {id}...");
                await APIRequest("trophies/add-achieved/", new string[] { $"username={UserName}", $"user_token={UserToken}", $"trophy_id={id}" });
                Debug.Log("Trophy granted.");
                OnTrophy?.Invoke(TrophyEventType.Grant, id);
            }

            public static async Task Revoke(int id)
            {
                RequireAuthenticated("Attempt to revoke trophy without an authenticated user.");
                Debug.Log($"Revoking trophy {id}...");
                await APIRequest("trophies/remove-achieved/", new string[] { $"username={UserName}", $"user_token={UserToken}", $"trophy_id={id}" });
                Debug.Log("Trophy revoked.");
                OnTrophy?.Invoke(TrophyEventType.Revoke, id);
            }

            public static async Task<TrophyData> Fetch(int id)
            {
                RequireAuthenticated("Attempt to fetch trophy data without an authenticated user.");
                Debug.Log($"Fetching trophy data for {id}...");
                JSONNode response = await APIRequest("trophies/", new string[] { $"username={UserName}", $"user_token={UserToken}", $"trophy_id={id}" });
                Debug.Log("Fetched trophy data.");
                TrophyData trophyData = new TrophyData(response["trophies"].AsArray[0]);
                return trophyData;
            }

            public static async Task<List<TrophyData>> List(bool all = true, bool achieved = true)
            {
                RequireAuthenticated("Attempt to list trophies without an authenticated user.");
                List<string> queries = new List<string>() { $"username={UserName}", $"user_token={UserToken}" };
                if (!all)
                {
                    queries.Add($"achieved={achieved.ToString().ToLower()}");
                }

                Debug.Log("Fetching trophy list...");
                JSONArray trophies = (await APIRequest("trophies/", queries.ToArray()))["trophies"].AsArray;
                List<TrophyData> trophyDatas = new List<TrophyData>();
                if (trophies != null)
                {
                    foreach (JSONObject trophy in trophies)
                    {
                        trophyDatas.Add(new TrophyData(trophy));
                    }
                }
                Debug.Log("Fetched trophy list.");
                return trophyDatas;
            }
        }
        

        public static async Task<DateTime> GetServerTime(bool localTime = true)
        {
            JSONNode response = await APIRequest("time/", null);
            if (localTime)
            {
                return DateTimeOffset.FromUnixTimeSeconds(response["timestamp"].AsLong).ToLocalTime().DateTime;
            }
            else
            {
                return DateTimeOffset.FromUnixTimeSeconds(response["timestamp"].AsLong).DateTime;
            }
        }

        private static void RequireAuthenticated(string message = "Not authenticated.")
        {
            if (!IsAuthenticated)
            {
                throw new AuthError(message);
            }
        }

        private static async Task<JSONNode> APIRequest(string endpoint, string[] queries, bool handleerrors = true)
        {
            string request = "";

            // Form a complete request
            switch (Version)
            {
                case APIVersion.v1_2:
                    request += "https://api.gamejolt.com/api/game/v1_2/";
                    break;
                default:
                    throw new APIError("Unsupported API version.");
            }
            request += $"{endpoint}?game_id={GameID}";
            if (queries != null)
            {
                foreach (string query in queries)
                {
                    request += $"&{query}";
                }
            }

            string result = (await AsyncWebRequest.GET(AddSignature(request))).downloadHandler.text;
            JSONNode response = JSON.Parse(result)["response"];
            if (response["success"] == "false" && handleerrors)
            {
                string message = response["message"];
                switch (message)
                {
                    case "No such user with the credentials passed in could be found.":
                        throw new AuthError(message);
                    default:
                        throw new APIError(message);
                }
            }
            
            return response;
        }

        private static string AddSignature(string strToSign)
        {
            return $"{strToSign}&signature={Helper.Md5Sum(strToSign + GameKey)}";
        }

        private static class Helper
        {
            public static string Md5Sum(string strToEncrypt)
            {
                System.Text.UTF8Encoding ue = new System.Text.UTF8Encoding();
                byte[] bytes = ue.GetBytes(strToEncrypt);

                // encrypt bytes
                System.Security.Cryptography.MD5CryptoServiceProvider md5 = new System.Security.Cryptography.MD5CryptoServiceProvider();
                byte[] hashBytes = md5.ComputeHash(bytes);

                // Convert the encrypted bytes back to a string (base 16)
                string hashString = "";

                for (int i = 0; i < hashBytes.Length; i++)
                {
                    hashString += System.Convert.ToString(hashBytes[i], 16).PadLeft(2, '0');
                }

                return hashString.PadLeft(32, '0');
            }
        }
    }
}


Runtime/GameJoltAPI.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Networking;
using DateTime = System.DateTime;
using SimpleJSON;

namespace WattanaGaming.GameJoltAPI
{
    public class GameJoltAPI : MonoBehaviour
    {
        static GameJoltAPI _instance;
        public static GameJoltAPI Instance
        {
            get
            {
                if (_instance == null)
                {
                    GameObject go = new GameObject("GameJolt API Instance");
                    go.AddComponent<GameJoltAPI>();
                }

                return _instance;
            }
        }

        public static string gameID;
        public static string gameKey;

        // User credentials are cached for later use after authentication.
        public static string username { get; private set; }
        public static string userToken { get; private set; }

        public static bool isAuthenticated { get; private set; }
        public static bool isAuthenticating { get; private set; }

        private static string baseURL = "https://api.gamejolt.com/api/game/v1_2/";

        /// <summary>
        /// Gets invoked upon an authentication attempt. Boolean indicates whether the attempt is successful or not.
        /// </summary>
        public event System.Action<bool> OnAuthenticate;
        /// <summary>
        /// Gets invoked upon granting or revoking a trophy. Int indicates trophy ID and TrophyEventType indicates the event type(Grant or Revoke).
        /// </summary>
        public event System.Action<int, TrophyEventType> OnTrophy;

        void Awake()
        {
            _instance = this;
        }

        /// <summary>
        /// Authenticate a GameJolt user with the specified username and token.
        /// </summary>
        /// <param name="name">The user's GJ username.</param>
        /// <param name="token">The user's GJ token.</param>
        /// <param name="callback">Optional callback.</param>
        /// <param name="forced">Force a re-authentication.</param>
        public void Authenticate(string name, string token, System.Action<bool> callback = null, bool forced = false)
        {
            if ((isAuthenticated && !forced) || isAuthenticating)
            {
                Debug.LogWarning("Already authenticated or is currently authenticating.");
                return;
            }
            isAuthenticating = true;
            Debug.Log($"Attempting to authenticate as {name}...");
            APIRequest("users/auth/", new List<string>() { $"username={name}", $"user_token={token}" }, (JSONNode response) =>
             {
                 username = name;
                 userToken = token;
                 isAuthenticated = true;
                 isAuthenticating = false;
                 callback?.Invoke(true);
                 OnAuthenticate?.Invoke(true);
             },
            () =>
            {
                // Debug.LogError($"Authentication failed. {response["message"]}");
                username = userToken = "";
                isAuthenticated = false;
                isAuthenticating = false;
                callback?.Invoke(false);
                OnAuthenticate?.Invoke(false);
                return;
            });
        }

        /// <summary>
        /// Grant the user a trophy with the specified ID.
        /// </summary>
        /// <param name="id"></param>
        public void GrantTrophy(int id)
        {
            if (!isAuthenticated)
            {
                Debug.LogError("Attempt to grant trophy without an authenticated user.");
                return;
            }
            Debug.Log($"Granting trophy {id}...");
            APIRequest("trophies/add-achieved/", new List<string>() { $"username={username}", $"user_token={userToken}", $"trophy_id={id}" }, (JSONNode _) =>
            {
                Debug.Log("Trophy granted.");
                OnTrophy?.Invoke(id, TrophyEventType.Grant);
            });
        }

        /// <summary>
        /// Revoke a trophy with the specified ID from the user.
        /// </summary>
        /// <param name="id"></param>
        public void RevokeTrophy(int id)
        {
            if (!isAuthenticated)
            {
                Debug.LogError("Attempt to revoke trophy without an authenticated user.");
                return;
            }
            Debug.Log($"Revoking trophy {id}...");
            APIRequest("trophies/remove-achieved/", new List<string>() { $"username={username}", $"user_token={userToken}", $"trophy_id={id}" }, (JSONNode _) =>
            {
                Debug.Log("Trophy revoked.");
                OnTrophy?.Invoke(id, TrophyEventType.Revoke);
            });
        }

        /// <summary>
        /// Get informations about a trophy.
        /// </summary>
        /// <param name="id">The ID of the trophy to fetch.</param>
        /// <param name="callback">Optional callback</param>
        public void FetchTrophy(int id, System.Action<TrophyData> callback=null)
        {
            if (!isAuthenticated)
            {
                Debug.LogError("Attempt to fetch trophy data without an authenticated user.");
                return;
            }
            Debug.Log($"Fetching trophy data for {id}...");
            APIRequest("trophies/", new List<string>() { $"username={username}", $"user_token={userToken}", $"trophy_id={id}" }, (JSONNode response) =>
            {
                Debug.Log("Fetched trophy data.");
                TrophyData trophyData = new TrophyData(response["trophies"].AsArray[0]);
                callback?.Invoke(trophyData);
            });
        }

        /// <summary>
        /// Get a list of trophies
        /// </summary>
        /// <param name="all">Get all of the trophies if set to true.</param>
        /// <param name="achieved">Only list achieved trophies if true and vice versa. Will be ignored if `all` is true.</param>
        /// <param name="callback">Optional callback.</param>
        public void ListTrophies(bool all, bool achieved, System.Action<List<TrophyData>> callback = null)
        {
            if (!isAuthenticated)
            {
                Debug.LogError("Attempt to list trophies without an authenticated user.");
                return;
            }
            List<string> queries = new List<string>() { $"username={username}", $"user_token={userToken}" };
            if (!all)
            {
                queries.Add($"achieved={achieved.ToString().ToLower()}");
            }
            Debug.Log("Fetching trophy list...");
            APIRequest("trophies/", queries, (JSONNode response) =>
            {
                JSONArray trophies = response["trophies"].AsArray;
                List<TrophyData> trophyDatas = new List<TrophyData>();
                if (trophies != null)
                {
                    foreach (JSONObject trophy in trophies)
                    {
                        trophyDatas.Add(new TrophyData(trophy));
                    }
                }
                Debug.Log("Fetched trophy list.");
                callback?.Invoke(trophyDatas);
            });
        }

        public void GetServerTime(System.Action<DateTime> callback = null)
        {
            APIRequest("time/", new List<string>(), (JSONNode response) =>
            {
                DateTime serverTime = new DateTime(response["year"].AsInt, response["month"].AsInt, response["day"].AsInt, response["hour"].AsInt, response["minute"].AsInt, response["second"].AsInt);
                callback?.Invoke(serverTime);
            });
        }

        void APIRequest(string endpoint, List<string> queries, System.Action<JSONNode> OnSuccess, System.Action OnError = null)
        {
            string request = $"{baseURL}{endpoint}?";
            request += $"game_id={gameID}";
            foreach (string query in queries)
            {
                request += $"&{query}";
            }
            StartCoroutine(GetRequest(AddSignature(request), (UnityWebRequest webRequest) =>
            {
                JSONNode response = JSON.Parse(webRequest.downloadHandler.text)["response"];
                if (response["success"] == "false")
                {
                    Debug.LogError($"Error: {response["message"]}");
                    OnError?.Invoke();
                    return;
                }
                OnSuccess?.Invoke(response);
            }));
        }

        private IEnumerator GetRequest(string uri, System.Action<UnityWebRequest> callback = null)
        {
            UnityWebRequest webRequest = UnityWebRequest.Get(uri);
            // Request and wait for the desired page.
            yield return webRequest.SendWebRequest();

            switch (webRequest.result)
            {
                case UnityWebRequest.Result.ConnectionError:
                    Debug.LogError("Connection error whilst making a web request.");
                    break;
                case UnityWebRequest.Result.DataProcessingError:
                    Debug.LogError("Data processing error whilst making a web request.");
                    break;
                case UnityWebRequest.Result.Success:
                    callback?.Invoke(webRequest);
                    break;
            }
        }

        private string AddSignature(string strToSign)
        {
            return $"{strToSign}&signature={Md5Sum(strToSign + gameKey)}";
        }
        
        private string Md5Sum(string strToEncrypt)
        {
            System.Text.UTF8Encoding ue = new System.Text.UTF8Encoding();
            byte[] bytes = ue.GetBytes(strToEncrypt);

            // encrypt bytes
            System.Security.Cryptography.MD5CryptoServiceProvider md5 = new System.Security.Cryptography.MD5CryptoServiceProvider();
            byte[] hashBytes = md5.ComputeHash(bytes);

            // Convert the encrypted bytes back to a string (base 16)
            string hashString = "";

            for (int i = 0; i < hashBytes.Length; i++)
            {
                hashString += System.Convert.ToString(hashBytes[i], 16).PadLeft(2, '0');
            }

            return hashString.PadLeft(32, '0');
        }
    }
}
